### DISCUSSION TITLE
Column chart - show value above each column

### DISCUSSION THREAD

**Discussion by venator85** [patrykandpatrick/vico#573] [Questions] [answered]
Hi, I'm trying to recreate the following design:

![Simulator Screenshot - iPhone 15 Pro - 2024-02-09 at 21 56 39](https://github.com/patrykandpatrick/vico/assets/479928/81cf6e87-aa7a-47ec-9eb9-cfd9bec10760)

in particular, I'm struggling to find a way to display the value above each column (it shows the value of the column itself). I'm using vico 2.0.0-alpha.6.

Can you help me?

Thanks

**Comment by patrickmichalik** [comment]
Hello! These are called data labels in Vico. Please let me know if you’re using Vico with Compose or with views so I can let you know how to add them to your chart.

**Reply by venator85** [reply]
Hi, I'm using Compose.
I found out myself, and I'm doing it like this:

    rememberColumnCartesianLayer(
        dataLabel = textComponent {
            textSizeSp = 10f
        },
        dataLabelVerticalPosition = VerticalPosition.Top,
        ...

However, the label is shown inside the column for the columns with the highest value (I think not to print them outside the chart area):

<img width="417" alt="image" src="https://github.com/patrykandpatrick/vico/assets/479928/d981c72d-aa91-4dc0-a201-03639cb34329">

Would it be possible to rescale down proportionately all the columns so that every data label is printed above (outside) the column?

Many thanks!

**Reply by Laul** [reply]
Hey ! 
did you try to just change the maxY of your chart ? Something like adding this to your rememberCartesianChart:: 
`axisValueOverrider = AxisValueOverrider.fixed(maxY = currentMax + someValue)`

**Reply by venator85** [reply]
Hi, yes I thought about that, but the value to add would depend on the magnitude of the data (and thus cannot be hardcoded), rather than on the font size of the label.

A better way would be to add a fixed "internal top padding" in dp above the chart area (where the data label can be drawn of course), so that I can hardcode a value that depends only on the font size of the data label. Can it be done?

**Reply by patrickmichalik** [reply, marked as answer]
I’m afraid this isn’t possible yet, @venator85, but we’ll be looking into adding such functionality. In the meantime, as mentioned by @Laul, `AxisValueOverrider` can be used. When the target values depend on the default ones, creating an `AxisValueOverrider` implementation directly is generally preferable to using `AxisValueOverrider.fixed`. This is particularly important when a `CartesianChartModelProducer` is used—see the second warning box [here](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.6/cartesian-charts/data).

As you’ve noted, a hard-coded value would be insufficient, but I’d say the reason is font scaling, not the fact that the data’s magnitude may change. You can resolve the magnitude problem by using a multiplier instead of adding a delta. Consider the following example. Whatever the data’s maximum _y_ value is, the top 10% of the `ColumnCartesianLayer` is left empty.

```kt
object : AxisValueOverrider<ColumnCartesianLayerModel> {
    override fun getMaxY(model: ColumnCartesianLayerModel) = 1.1f * model.maxY
}
```

If font size were expressed in dp, 10% of the `ColumnCartesianLayer`’s height would always be the same number of dp, and you’d be able to find a constant multiplier for which the data labels aren’t shifted. However, font size is measured in sp, and 1 sp corresponds to a different number of dp depending on the user’s font scale. As a result, even if the chart’s height is always the same number of dp, the following are true:

- The height of the `ColumnCartesianLayer` changes. This is because the size of the `HorizontalAxis` labels is expressed in sp. The labels’ height influences the `HorizontalAxis`’s height, and as the `HorizontalAxis` grows, the `ColumnCartesianLayer` shrinks.
- The space needed for the data labels (measured in dp) changes.
- These two values’ ratio changes.

Thus, unless the chart’s height is expressed in sp, which I imagine you’d like to avoid, the `ColumnCartesianLayerModel#maxY` multiplier will have to be calculated at runtime based on the font scale.

As a final note, I wanted to let you know that the `textComponent` function doesn’t use `remember` and isn’t intended for Compose. For better performance, consider using `rememberTextComponent`.

Updates regarding the ability to add top padding to the `CartesianLayer` area will be shared here: #448. If you have any questions, please let me know. Thanks for the help, @Laul!

**Reply by venator85** [reply]
Thanks for your detailed answer! Multiplying the max by an appropriate value looks for now, anyway I'll keep an eye on #448.

**Reply by patrickmichalik** [reply]
I’m glad to hear that, @venator85. You’re welcome!
